sdk/attestation/azure-security-attestation/inc/azure/attestation/attestation_client.hpp

```diff
@@ -0,0 +1,410 @@
+   * is as a binary blob of data. That is needed if either the data sent by the enclave isn't a JSON
+   * object - for instance, if the RunTime data contained an asymmetric key which is formatted as a
+   * PEM encoded key, it should be interpreted as a binary blob
+   *
+   * If you ask for the RunTime data to be included in the token as binary, then it will be
+   * base64url encoded in the "x-ms-maa-enclavehelddata" claim in the output token (the
+   * {@link AttestationResult#getEnclaveHeldData()} property).
+   *
+   * If you ask for the RunTime data to be included in the token as JSON, then it will be included
+   * in the "x-ms-maa-runtimeClaims" claim in the output token (the {@link
+   * AttestationResult#getRuntimeClaims()} property).
+   *
+   * In addition to the Attest APIs, the {@link AttestationClient} object also contains helper APIs
+   * which can be used to retrieve the OpenId Metadata document and signing keys from the service.
+   *
+   *
+   * The OpenId Metadata document contains properties which describe the attestation service.
+   *
+   * The Attestation Signing Keys describe the keys which will be used to sign tokens generated by
+   * the attestation service. All tokens emitted by the attestation service will be signed by one
+   * of the certificates listed in the attestation signing keys.
+   *
+   */
+
+  class AttestationClient final {
+  private:
+    Azure::Core::Url m_endpoint;
+    std::string m_apiVersion;
+    std::shared_ptr<Azure::Core::Http::_internal::HttpPipeline> m_pipeline;
+    AttestationTokenValidationOptions m_tokenValidationOptions;
+
+    mutable std::shared_timed_mutex m_sharedStateLock;
+    mutable std::vector<Models::AttestationSigner> m_attestationSigners;
+
+    void CacheAttestationSigners(Azure::Core::Context const& context) const;
+
+  public:
+    /**
+     * @brief Destructor.
+     *
+     */
+    virtual ~AttestationClient() = default;
+
+    /**
+     * @brief Construct a new Attestation Client object
+     *
+     * @param endpoint The URL address where the client will send the requests to.
+     * @param credential OPTIONAL The authentication method to use (required for TPM attestation).
+     * @param options The options to customize the client behavior.
+     */
+    explicit AttestationClient(
+        std::string const& endpoint,
+        std::shared_ptr<Core::Credentials::TokenCredential const> credential,
+        AttestationClientOptions options = AttestationClientOptions());
+
+    explicit AttestationClient(
+        std::string const& endpoint,
+        AttestationClientOptions options = AttestationClientOptions())
+        : AttestationClient(endpoint, nullptr, options)
+    {
+    }
+
+    /**
+     * @brief Returns the version the client was configured with.
+     *
+     * @returns The API version used when communicating with the attestation service.
+     */
+    std::string ClientVersion() const;
```

Comment:
nit. return a `std::string const&` instead of `std::string` to let caller decide if they want to make a copy or just read the value.